jetty-ee11/jetty-ee11-servlet/src/main/java/org/eclipse/jetty/ee11/servlet/ServletContextHolder.java

```diff
@@ -0,0 +1,163 @@
+
+import jakarta.servlet.ServletContext;
+import jakarta.servlet.UnavailableException;
+import org.eclipse.jetty.ee.BaseHolder;
+import org.eclipse.jetty.ee.Source;
+import org.eclipse.jetty.server.handler.ContextHandler;
+import org.eclipse.jetty.util.annotation.ManagedAttribute;
+import org.eclipse.jetty.util.annotation.ManagedObject;
+
+/**
+ * Specialization of BaseHolder for servlet-related classes
+ *
+ * @param <T> the type of holder
+ */
+@ManagedObject("Holder - a container for servlets and the like")
+public abstract class ServletContextHolder<T> extends BaseHolder<T>
+{
+    private ServletHandler _servletHandler;
+    private String _displayName;
+    private String _name;
+
+    protected ServletContextHolder(Source source)
+    {
+        super(source);
+    }
+
+    @Override
+    public ContextHandler getContextHandler()
+    {
+        ContextHandler contextHandler = super.getContextHandler();
+        if (contextHandler == null && _servletHandler != null && _servletHandler.getServletContextHandler() != null)
+        {
+            contextHandler = _servletHandler.getServletContextHandler();
+            super.setContextHandler(contextHandler);
+        }
+        return contextHandler;
+    }
+
+    public ServletContext getServletContext()
+    {
+        if (_servletHandler != null)
+            return _servletHandler.getServletContext();
+        if (getContextHandler() != null && getContextHandler().getContext() instanceof ServletContextHandler.ServletScopedContext servletScopedContext)
+            return servletScopedContext.getServletContext();
+        return null;
+    }
+
+    public ServletContextHandler getServletContextHandler()
+    {
+        if (_servletHandler != null && _servletHandler.getServletContextHandler() != null)
+            return _servletHandler.getServletContextHandler();
+        if (getContextHandler() != null && getContextHandler().getContext() instanceof ServletContextHandler.ServletScopedContext servletScopedContext)
+            return servletScopedContext.getServletContextHandler();
+        return null;
+    }
+
+    public ServletHandler getServletHandler()
+    {
+        if (_servletHandler != null)
+            return _servletHandler;
+        if (getContextHandler() != null && getContextHandler().getContext() instanceof ServletContextHandler.ServletScopedContext servletScopedContext)
+            return servletScopedContext.getServletContextHandler().getServletHandler();
+        return null;
+    }
+
+    public void setServletHandler(ServletHandler handler)
+    {
+        _servletHandler = handler;
+        if (_servletHandler.getServletContextHandler() != null)
+            setContextHandler(_servletHandler.getServletContextHandler());
+    }
+
+    @Override
+    protected Exception newUnavailableException(String message)
+    {
+        return new UnavailableException(message);
+    }
+
+    @ManagedAttribute(value = "Display Name", readonly = true)
```

Comment:
Can't all the methods from here on down be pushed down into `BaseHolder`? They're independent of the servlet spec version, and I don't think its important if they are not used by some of the derived classes.